tests/xdc/nexus_request_forwarding_test.go

```diff
@@ -0,0 +1,339 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package xdc
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	enumspb "go.temporal.io/api/enums/v1"
+	nexuspb "go.temporal.io/api/nexus/v1"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/metrics/metricstest"
+	"go.temporal.io/server/common/namespace"
+	cnexus "go.temporal.io/server/common/nexus"
+	"go.temporal.io/server/components/nexusoperations"
+	"go.temporal.io/server/tests"
+)
+
+var op = nexus.NewOperationReference[string, string]("my-operation")
+
+type NexusRequestForwardingSuite struct {
+	xdcBaseSuite
+}
+
+func TestNexusRequestForwardingTestSuite(t *testing.T) {
+	flag.Parse()
+	suite.Run(t, new(NexusRequestForwardingSuite))
+}
+
+func (s *NexusRequestForwardingSuite) SetupSuite() {
+	s.dynamicConfigOverrides = map[dynamicconfig.Key]interface{}{
+		// Make sure we don't hit the rate limiter in tests
+		dynamicconfig.FrontendGlobalNamespaceNamespaceReplicationInducingAPIsRPS: 1000,
+		dynamicconfig.FrontendEnableNexusAPIs:                                    true,
+		nexusoperations.Enabled:                                                  true,
+		dynamicconfig.FrontendRefreshNexusIncomingServicesMinWait:                5 * time.Millisecond,
+	}
+	s.setupSuite([]string{"nexus_request_forwarding_active", "nexus_request_forwarding_standby"})
+}
+
+func (s *NexusRequestForwardingSuite) SetupTest() {
+	s.setupTest()
+}
+
+func (s *NexusRequestForwardingSuite) TearDownSuite() {
+	s.tearDownSuite()
+}
+
+// Only tests dispatch by namespace+task_queue.
+// TODO: Add test cases for dispatch by incoming service ID once incoming services support replication.
+func (s *NexusRequestForwardingSuite) TestStartOperationForwardedFromStandbyToActive() {
+	ns := s.createNexusRequestForwardingNamespace()
+
+	testCases := []struct {
+		name      string
+		taskQueue string
+		handler   func(*workflowservice.PollNexusTaskQueueResponse) (*nexuspb.Response, *nexuspb.HandlerError)
+		assertion func(*testing.T, *nexus.ClientStartOperationResult[string], error, map[string][]*metricstest.CapturedRecording, map[string][]*metricstest.CapturedRecording)
+	}{
+		{
+			name:      "success",
+			taskQueue: fmt.Sprintf("%v-%v", "test-task-queue", uuid.New()),
+			handler: func(res *workflowservice.PollNexusTaskQueueResponse) (*nexuspb.Response, *nexuspb.HandlerError) {
+				s.Equal("true", res.Request.Header["xdc-redirection-api"])
```

Comment:
I'm debating whether this header should make it all the way to the worker. Users may find it surprising.
Let's track this for consideration. Not blocking this PR.

Reply:
Still not blocking the PR but I think we should remove this header before it hits the worker.